### PR TITLE
Recipe app

### PR DESCRIPTION
**Pull request by svitalis123**
# Recipe app
> Team
- [@svitalis123 ]
- [@chuaindia ]

## we accomplished

- [ ] Created a login page.
- [ ] Created a registration page.
- [ ] Created a food list (with all CRUD implementation, except for 'update').
- [ ] Created a recipes list (with all CRUD implementation, except for 'update').
- [ ] Created a general shopping list view (all missing food for all your recipes and total price).
- [ ] Created a list of all public recipes from other users with their names and total prices.

### CODE REVIEW

File: app/views/foods/_food.html.erb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+  <p>
+    <%= food.measurement_unit %>
+  </p>
+
+  <p>
+    $<%= food.price %>
+  </p>
+
+</div>
```
Comment (Nemwel-Boniface) [inline app/views/foods/_food.html.erb]: - [ ] You have done a good job listing the foods in a tabular form as it was required as per the provided ERD. Kindly add the delete food button which is currently missing in your design but it supposed to be according to the project requirements.

**This is how your design looks like:**

image


**This is the desired design with the delete button:**

image

File: app/models/ability.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    user ||= User.new
+    can :read, Recipe, user_id: user.id
+    can :manage, Recipe, user_id: user.id
```
Comment (Nemwel-Boniface) [inline app/models/ability.rb]: - [x] Kindly add authorization rules that only allow the owner of food to manage it. A food should only be able to be deleted by the owner who created it alone.

image

File: app/views/general_shopping_list/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            <%@missing_ingredients.each do |ingredient| %>
+              <%= render "ingredient", ingredient: ingredient %>
+            <% end %>
+          </tbody>
+        </table>
+      </div>
+    </div>
+  </div>
+</div>
+
```
Comment (Nemwel-Boniface) [inline app/views/general_shopping_list/index.html.erb]: - [ ] For the general shopping list, you are required to implement the following:

1. Should count the total food items and total price of the missing food
2. Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user)

image

**This is how your general shopping list is looking like:**

image

**This is how it is supposed to look like:**

image

File: db/schema.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+ActiveRecord::Schema[7.0].define(version: 2023_02_16_083549) do
+  # These are extensions that must be enabled in order to support this database
+  enable_extension "plpgsql"
+
+  create_table "food_recipes", id: false, force: :cascade do |t|
+    t.bigint "recipe_id", null: false
+    t.bigint "food_id", null: false
+    t.integer "quantity"
+  end
```
Comment (Nemwel-Boniface) [inline db/schema.rb]: - [ ] According to the project ERD, the recipe_id and food_id should also have foreign key references to the respective tables. Kindly include them as foreign key indexes.

image

File: spec/controllers/recipe_spec.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  context 'should render_template' do
+    it 'index action should render index template' do
+      get '/users/1/recipes'
+      expect(response).to render_template('index')
+    end
+    it 'show action to render template show' do
+      get '/users/1/recipes/1'
+      expect(response).to render_template('show')
+    end
+  end
```
Comment (Nemwel-Boniface) [inline spec/controllers/recipe_spec.rb]: - [ ] For some reason, the following test is failing. Below is the error I am getting in the console. Kindly look into why this is the case.

image

File: spec/models/recipe_spec.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    end
+    it 'cooking_time to be two hours' do
+      expect(subject.cooking_time).to eq(2)
+    end
+    it 'cooking_time above 400 to be not valid' do
+      subject.cooking_time = 401
+      expect(subject).to_not be_valid
+    end
+  end
+end
```
Comment (Nemwel-Boniface) [inline spec/models/recipe_spec.rb:46]: - [ ] For some reason, the following test is failing. Below is the error I am getting in the console. Kindly look into why this is the case.

image

File: spec/views/recipe_index_spec.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  end
+  context 'should render index page accordingly' do
+    it 'have the required content' do
+      visit "/users/#{@user.id}/recipes/"
+      expect(page).to have_content(@recipe.name)
+      expect(page).to have_content(@recipe.description)
+      expect(page).to have_button('Remove')
+    end
+  end
+end
```
Comment (Nemwel-Boniface) [inline spec/views/recipe_index_spec.rb]: - [x] For some reason, the following test is failing. Below is the error I am getting in the console. Kindly look into why this is the case.

image

File: spec/views/recipe_show_spec.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    it 'have the required content' do
+      visit "/users/#{@user.id}/recipes/#{@recipe.id}"
+      expect(page).to have_content(@recipe.name)
+      expect(page).to have_content(@recipe.preparation_time)
+      expect(page).to have_content(@recipe.cooking_time)
+      expect(page).to have_button('Generate Shopping List')
+      expect(page).to have_button('Add Ingredient')
+    end
+  end
+end
```
Comment (Nemwel-Boniface) [inline spec/views/recipe_show_spec.rb]: - [ ] For some reason, the following test is failing. Below is the error I am getting in the console. Kindly look into why this is the case.

image

**Review by Nemwel-Boniface — CHANGES_REQUESTED** [review]
## Hello @svitalis123  and @chuaindia      :wave: , 
I'm Nemwel your reviewer for this PR :man_beard: 

You both have an amazing Recipe project that I enjoyed reviewing! :clap: 
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights

Great job setting up the correct linters ✔️ 
Professional documentation ✔️ 
Correct workflow ✔️ 
Meaningful commit messages ✔️  


## ♻️ Required Changes ♻️

_Check the comments under the review._

## [Optional] suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding! :partying_face: 👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me **@Nemwel-Boniface** in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/recipe_index_spec.rb (hunk 7 above)
Comment (svitalis123) [reply spec/views/recipe_index_spec.rb]: resolved

File: spec/models/recipe_spec.rb (hunk 6 above)
Comment (svitalis123) [reply spec/models/recipe_spec.rb:46]: resolved

File: db/schema.rb (hunk 4 above)
Comment (svitalis123) [reply db/schema.rb]: resolved

File: app/views/general_shopping_list/index.html.erb (hunk 3 above)
Comment (svitalis123) [reply app/views/general_shopping_list/index.html.erb]: thank you did

File: app/views/foods/_food.html.erb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  </p>
+
+  <p>
+    <%= food.measurement_unit %>
+  </p>
+
+  <p>
+    $<%= food.price %>
+  </p>
+  <%= button_to "delete", user_foods_path(food), method: :delete %>
```
Comment (victorokoroji) [inline app/views/foods/_food.html.erb]: - [ ] Good job adding the delete functionality as suggested by the previous reviewer 👍🏽 . The only challenge left is that the button is not deleting the food from the list. KIndly check it out

image

**Review by victorokoroji — CHANGES_REQUESTED** [review]
Hi @svitalis123, 

Great work on making the changes requested by a previous reviewer 👏 🔥 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



## Top highlight 🎯

- The PR title and description are professional and descriptive ✅ 
- The commit messages are meaningful ✅ 
- Linters are working properly ✅ 


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/foods/_food.html.erb (hunk 9 above)
Comment (svitalis123) [reply app/views/foods/_food.html.erb]: i did that thank you for the review

**Review by rustam-tolipov — APPROVED** [review]
Hi @svitalis123, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Developer-512px-10

## Highlights
1. No Linters errors 💯
2. All requirements are complete 💯

## Optional suggestions
No optional suggestions :)

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**